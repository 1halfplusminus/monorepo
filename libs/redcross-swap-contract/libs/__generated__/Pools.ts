/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: Pools
// ====================================================

export interface Pools_pools_token0 {
  __typename?: 'Token';
  id: string;
  symbol: string;
  decimals: any;
}

export interface Pools_pools_token1 {
  __typename?: 'Token';
  id: string;
  symbol: string;
  decimals: any;
}

export interface Pools_pools {
  __typename?: 'Pool';
  id: string;
  token0: Pools_pools_token0;
  token1: Pools_pools_token1;
  feeTier: any;
  token0Price: any;
  token1Price: any;
}

export interface Pools {
  pools: Pools_pools[];
}
