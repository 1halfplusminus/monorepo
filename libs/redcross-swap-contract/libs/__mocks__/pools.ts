import poolJson from './pools.json';

export const QUERY_POOLS_RESULT = poolJson;
