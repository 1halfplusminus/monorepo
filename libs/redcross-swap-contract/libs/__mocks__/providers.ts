const result = require('dotenv').config();

import { ethers } from 'ethers';
import { SupportedChainId } from '../constants/chains';

const rinkeby = {
  url: process.env.RINKEBY_URL,
  blockGasLimit: 9500000,
  accounts: {
    mnemonic: process.env.RINKEBY_MNEMONIC,
  },
};

export const chainId = SupportedChainId.RINKEBY;
const _provider = new ethers.providers.JsonRpcProvider(rinkeby.url, chainId);
/* let wallet = ethers.Wallet.fromMnemonic(rinkeby.accounts.mnemonic); */
export const provider = _provider;
export const signer = provider.getSigner();
