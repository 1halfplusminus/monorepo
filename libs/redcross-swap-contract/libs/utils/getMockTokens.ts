import { tokenList } from '../__mocks__';
import { record as R, nonEmptyArray as NEA } from 'fp-ts';
import { pipe } from 'fp-ts/function';
import { filterByChainId } from '../index';

export const getMockTokens = (chaindId: number = 1) =>
  pipe(
    tokenList,
    filterByChainId(chaindId),
    NEA.groupBy((t) => t.symbol),
    R.map((v) => v[0])
  );
