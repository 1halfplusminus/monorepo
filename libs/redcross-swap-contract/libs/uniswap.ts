import {
  computePoolAddress,
  Pool,
  Tick,
  TickListDataProvider,
  tickToPrice,
} from '@uniswap/v3-sdk';
import { IUniswapV3Pool } from '../typechain/IUniswapV3Pool';
import { IUniswapV3Pool__factory } from '../typechain/factories/IUniswapV3Pool__factory';
import { Provider } from '@ethersproject/providers';
import { Address } from 'hardhat-deploy/dist/types';
import { BigNumber, ethers } from 'ethers';
import { Token as UToken } from '@uniswap/sdk-core';
import { constant, flow, pipe } from 'fp-ts/function';
import {
  task as T,
  option as O,
  array as A,
  taskOption as TO,
  taskEither as TE,
  either as E,
  record as R,
  function as F,
} from 'fp-ts';
import { FeeAmount } from '../test/shared/constants';
import type { Option } from 'fp-ts/Option';
import { Token } from '.';
import { useMemo } from 'react';
import { sequenceT } from 'fp-ts/lib/Apply';
import { useEffect, useState } from 'react';
import { useCallback } from 'react';
import {
  usePool,
  PoolsList,
  poolSelector,
  eqPoolTokenToken,
} from './uniswap-subgraph';
import { useIsMounted } from './index';
import { IQuoter__factory } from '../typechain/factories/IQuoter__factory';
import {
  Pools_pools,
  Pools_pools_token0,
  Pools_pools_token1,
} from './__generated__/Pools';

interface Immutables {
  factory: Address;
  token0: Address;
  token1: Address;
  fee: number;
  tickSpacing: number;
  maxLiquidityPerTick: BigNumber;
}
interface State {
  liquidity: BigNumber;
  sqrtPriceX96: BigNumber;
  tick: number;
  observationIndex: number;
  observationCardinality: number;
  observationCardinalityNext: number;
  feeProtocol: number;
  unlocked: boolean;
}
export const poolAddress = '0x8ad599c3A0ff1De082011EFDDc58f1908eb6e6D8';
export interface UseQuoter {
  provider: Option<Provider>;
}
export const useQuoter = ({ provider }: UseQuoter) => {
  const quoter = useMemo(
    () =>
      pipe(
        provider,
        O.map((p) =>
          IQuoter__factory.connect(
            '0xb27308f9F90D607463bb33eA1BeBb41C27CE5AB6',
            p
          )
        )
      ),
    [provider]
  );
  return quoter;
};
export const createPoolContract = (provider: Provider) => (
  address: string = poolAddress
) => IUniswapV3Pool__factory.connect(address, provider);
export const createPoolContractFromToken = (provider: Provider) => (
  tokenA: UToken,
  tokenB: UToken,
  feeAmount: FeeAmount,
  factoryAddress: Address
) =>
  pipe(
    computePoolAddress({
      factoryAddress: factoryAddress,
      fee: feeAmount,
      tokenA: tokenA,
      tokenB: tokenB,
    }),
    createPoolContract(provider)
  );
export const createPoolContractFromOptionToken = (
  provider: Option<Provider>
) => (tokenA: Option<UToken>, tokenB: Option<UToken>, feeAmount: FeeAmount) =>
  pipe(
    provider,
    O.map((provider) =>
      pipe(
        tokenA,
        O.map((tokenA) =>
          pipe(
            tokenB,
            O.map((tokenB) =>
              createPoolContractFromToken(provider)(
                tokenA,
                tokenB,
                feeAmount,
                ''
              )
            )
          )
        )
      )
    ),
    O.flatten,
    O.flatten
  );

export async function getPoolImmutables(poolContract: IUniswapV3Pool) {
  const [
    factory,
    token0,
    token1,
    fee,
    tickSpacing,
    maxLiquidityPerTick,
  ] = await Promise.all([
    await poolContract.factory(),
    await poolContract.token0(),
    await poolContract.token1(),
    await poolContract.fee(),
    await poolContract.tickSpacing(),
    await poolContract.maxLiquidityPerTick(),
  ]);
  const poolImmutables: Immutables = {
    factory,
    token0,
    token1,
    fee,
    tickSpacing,
    maxLiquidityPerTick,
  };
  return poolImmutables;
}
export async function getPoolState(poolContract: IUniswapV3Pool) {
  const slot = await poolContract.slot0();
  const PoolState: State = {
    liquidity: await poolContract.liquidity(),
    sqrtPriceX96: slot[0],
    tick: slot[1],
    observationIndex: slot[2],
    observationCardinality: slot[3],
    observationCardinalityNext: slot[4],
    feeProtocol: slot[5],
    unlocked: slot[6],
  };
  return PoolState;
}
export const createUniswapToken = (token: Token) =>
  new UToken(
    token.chainId,
    token.address,
    token.decimals,
    poolSelector(token),
    token.fullName
  );

export const createUniswapTokenFromOption = (token: Option<Token>) =>
  pipe(
    token ? token : O.none,
    O.map((token) => createUniswapToken(token))
  );
export const getPrice = (baseToken: Token, quoteToken: Token, tick: number) =>
  tickToPrice(
    createUniswapToken(baseToken),
    createUniswapToken(quoteToken),
    tick
  );
export const createTokenFromSubgraph = (chainId: number) => (
  token: Pools_pools_token0 | Pools_pools_token1
) => pipe(new UToken(chainId, token.id, Number(token.decimals), token.symbol));

export const createPoolFromSubgrap = (chainId: number) => (
  pool: Pools_pools,
  tickSpacing: number,
  ticks: Tick[]
) =>
  pipe(
    new Pool(
      createTokenFromSubgraph(chainId)(pool.token0),
      createTokenFromSubgraph(chainId)(pool.token1),
      Number(pool.feeTier),
      Number(pool.sqrtPrice),
      Number(pool.liquidity),
      Number(pool.tick),
      new TickListDataProvider(ticks, tickSpacing)
    )
  );
export const createPool = (
  poolContract: IUniswapV3Pool,
  baseToken: UToken,
  quoteToken: UToken,
  feeAmount: FeeAmount
) =>
  pipe(
    flow(constant(poolContract), getPoolState),
    T.map(
      (state) =>
        new Pool(
          baseToken,
          quoteToken,
          feeAmount,
          state.sqrtPriceX96.toString(),
          state.liquidity.toString(),
          state.tick
        )
    )
  )();
const eqTokenUToken = (token: Token) => (uToken: UToken) =>
  token.address.toUpperCase() == uToken.address.toUpperCase();
export interface UseUniswapProps {
  tokenA: Option<Token>;
  tokenB: Option<Token>;
  provider: Option<Provider>;
  pools: Option<PoolsList>;
  fetchPoolsImmutable?: (contrat: IUniswapV3Pool) => Promise<Immutables>;
}
export const useUniswapPools = () => {};
export const useUniswap = ({
  tokenA,
  tokenB,
  provider,
  fetchPoolsImmutable = getPoolImmutables,
  pools,
}: UseUniswapProps) => {
  const isMounted = useIsMounted();
  const [pool, setPool] = useState<Option<Pool>>(O.none);
  const [poolImmutables, setPoolImmutables] = useState<Option<Immutables>>(
    O.none
  );
  const tokenAUniswap = useMemo(() => createUniswapTokenFromOption(tokenA), [
    tokenA,
  ]);
  const tokenBUniswap = useMemo(() => createUniswapTokenFromOption(tokenB), [
    tokenB,
  ]);
  const { pool: poolInfo } = usePool({ tokenA, tokenB, pools });

  const poolContract = useMemo(
    () =>
      pipe(
        O.some(createPoolContract),
        O.ap(provider),
        O.ap(
          pipe(
            poolInfo,
            O.map((p) => p.id)
          )
        )
      ),
    [poolInfo, provider]
  );
  useEffect(() => {
    pipe(
      poolContract,
      TO.fromOption,
      TO.chain((p) => TO.tryCatch(() => fetchPoolsImmutable(p))),
      TO.chain((r) => async () =>
        pipe(
          r,
          O.fromPredicate(() => isMounted.current)
        )
      ),
      TO.map((r) => setPoolImmutables(O.some(r)))
    )();
    return () => {
      setPoolImmutables(O.none);
    };
  }, [poolContract]);
  useEffect(() => {
    pipe(
      sequenceT(O.Apply)(poolContract, tokenAUniswap, tokenBUniswap, poolInfo),
      TO.fromOption,
      TO.chain(([poolContract, tokenA, tokenB, poolInfo]) =>
        TO.tryCatch(() =>
          createPool(poolContract, tokenA, tokenB, Number(poolInfo.feeTier))
        )
      ),
      TO.chain((r) => async () =>
        pipe(
          r,
          O.fromPredicate(() => isMounted.current)
        )
      ),
      TO.map((p) => setPool(O.some(p)))
    )();
    return () => {
      setPool(O.none);
    };
  }, [poolContract, tokenAUniswap, tokenBUniswap, poolInfo]);
  const getTokenPrice = useCallback(
    async (token: Option<Token> | Token) =>
      pipe(
        '_tag' in token ? token : O.some(token),
        (t) => sequenceT(O.Apply)(tokenAUniswap, tokenBUniswap, t, poolInfo),
        O.chain(([tokenAUniswap, tokenBUniswap, token, poolInfo]) =>
          pipe(
            eqPoolTokenToken(poolInfo.token0, token)
              ? O.some(poolInfo.token0Price)
              : eqPoolTokenToken(poolInfo.token1, token)
              ? O.some(poolInfo.token1Price)
              : O.none,
            O.map((t) => {
              console.log(t);
              return ethers.FixedNumber.fromString(
                Number(t).toFixed(18),
                'fixed128x18'
              ).toString();
            })
          )
        )
      ),
    [pool, tokenAUniswap, tokenBUniswap, poolInfo]
  );

  return {
    pool,
    tokenAUniswap,
    tokenBUniswap,
    poolContract,
    getTokenPrice,
    poolImmutables,
    poolInfo,
  };
};
