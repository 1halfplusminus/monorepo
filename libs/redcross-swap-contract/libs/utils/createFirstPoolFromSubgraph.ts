import { getMockTokens } from '../utils/getMockTokens';
import { option as O, record as R, function as F, array as A } from 'fp-ts';
import { defaultPools, groupBySymbol } from '../uniswap-subgraph';

import { createPoolFromSubgrap } from '../uniswap';
import { Pools_pools } from '../__generated__/Pools';
import { TokenListItem } from '../index';
import type { NonEmptyArray } from 'fp-ts/NonEmptyArray';
import { selectFirstPool } from '../utils/selectFirstPool';

export const createFirstPoolFromSubgrap = (
  tokenA: TokenListItem,
  tokenB: TokenListItem
) =>
  F.flow(
    (pools: Record<string, NonEmptyArray<Pools_pools>>) =>
      selectFirstPool(tokenA, tokenB)(pools),
    O.map((p) => createPoolFromSubgrap(1)(p, 1, []))
  );
