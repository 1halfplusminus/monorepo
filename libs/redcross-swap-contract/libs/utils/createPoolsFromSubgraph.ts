import { Tick } from '@uniswap/v3-sdk';
import {
  record as R,
  function as F,
  array as A,
  number as N,
  option as O,
} from 'fp-ts';
import { pipe } from 'fp-ts/function';
import { createPoolFromSubgrap } from '../uniswap';
import { PoolsList } from '../uniswap-subgraph';
import { contramap } from 'fp-ts/Ord';
import { Pools_pools_ticks } from '../__generated__/Pools';
import { TICK_SPACINGS } from '../../test/shared/constants';
const sortTickByIndex = pipe(
  N.Ord,
  contramap((t: Tick) => t.index)
);
const createTicksFromSubGrap = (t: Pools_pools_ticks) => {
  return new Tick({
    liquidityGross: t.liquidityGross,
    liquidityNet: t.liquidityNet,
    index: Number(t.id.split('#')[1]),
  });
};

export const createPoolsFromSubgraph = (chainId: number) => (
  pools: PoolsList
) =>
  pipe(
    pools,
    R.toArray,
    A.map(([key, p]) =>
      pipe(
        p,
        A.map((p) =>
          O.tryCatch(() =>
            createPoolFromSubgrap(chainId)(
              p,
              TICK_SPACINGS[p.feeTier],
              pipe(
                p.ticks,
                A.map(createTicksFromSubGrap),
                A.sort(sortTickByIndex)
              )
            )
          )
        )
      )
    ),
    A.flatten,
    A.compact
  );
