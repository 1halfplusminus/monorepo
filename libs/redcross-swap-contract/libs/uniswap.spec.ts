import {
  createUniswapToken,
  createPool,
  UseUniswapProps,
  useUniswap,
} from './uniswap';
import { pipe } from 'fp-ts/function';
import {
  array as A,
  record as R,
  option as O,
  taskOption as TO,
  function as F,
  task as T,
  number as N,
} from 'fp-ts';
import { FeeAmount } from '../test/shared/constants';
import {
  createPoolContractFromToken,
  createPoolFromSubgrap,
  useUniswapTokens,
} from './uniswap';
import { renderHook, act } from '@testing-library/react-hooks';
import { Pools_pools } from './__generated__/Pools';
import {
  defaultFetchPools,
  defaultPools,
  groupBySymbol,
  useFetchMore,
  UseFetchMore,
  usePools,
} from './uniswap-subgraph';
import { FACTORY_ADDRESS, Tick, Route } from '@uniswap/v3-sdk';
import { sequenceT } from 'fp-ts/Apply';
import { createPoolContract, useQuoter, useUniswapRoute } from './uniswap';
import { contramap } from 'fp-ts/Ord';

import { chainId, provider } from './__mocks__/providers';

import fetch from 'node-fetch';
import { Token } from '@uniswap/sdk-core';
import { getMockTokens } from './utils/getMockTokens';
import { createPoolsFromSubgraph } from './utils/createPoolsFromSubgraph';

(global as any).fetch = fetch as any;
jest.setTimeout(100000);

describe('Uniswap Lib', () => {
  it('it should fetch price correctly', async () => {
    const tokens = getMockTokens();
    const eth = tokens['BUSD'];
    const dai = tokens['WETH'];

    const tokenA = createUniswapToken(eth);
    const tokenB = createUniswapToken(dai);
    const poolContact = createPoolContractFromToken(provider)(
      tokenB,
      tokenA,
      FeeAmount.MEDIUM,
      FACTORY_ADDRESS
    );
    const pool = await createPool(poolContact, tokenA, tokenB, FeeAmount.LOW);
    console.log(pool.priceOf(tokenB).toFixed());
    /* const trade = Trade.bestTradeExactIn([pool], tokenA, tokenB, currencyOut); */
  });
});

describe('Use uniswap hook', () => {
  const tokens = getMockTokens();
  const tokenA = pipe(tokens, R.lookup('WETH'));
  const tokenB = pipe(tokens, R.lookup('DAI'));
  const useUniswapProps = {
    chainId: O.some(1),
    tokenA: tokenA,
    tokenB: tokenB,
    provider: O.some(provider),
    pools: O.none,
  } as UseUniswapProps;
  it('should fetch more correctly', async () => {
    const fetchMoreProps: UseFetchMore<Pools_pools> = {
      fetchMore: defaultFetchPools,
      first: 0,
      skip: 0,
    };
    const { result, waitForValueToChange } = renderHook(() =>
      useFetchMore(fetchMoreProps)
    );
    await act(async () => {
      let fetchResult = await result.current.fetchMore(1000);
      expect(fetchResult.length == 1000).toBe(true);
      expect(result.current.first).toEqual(1000);
      expect(result.current.skip).toEqual(0);
      expect(result.current.hasMore).toEqual(O.some(true));

      fetchResult = await result.current.fetchMore(500);

      expect(result.current.first).toEqual(500);
      expect(result.current.skip).toEqual(1000);
      expect(result.current.hasMore).toEqual(O.some(true));

      fetchResult = await result.current.fetchMore(500);
      expect(result.current.hasMore).toEqual(O.none);
      /* expect(fetchResult.length == 500).toBe(true);
      expect(result.current.first).toEqual(500);
      expect(result.current.skip).toEqual(1500); */
    });
  });
  it('it should fetch default pools correctly', async () => {
    expect((await defaultFetchPools(0, 50)).length).toEqual(50);
    expect((await defaultFetchPools(10, 40)).length).toEqual(40);
    expect((await defaultFetchPools(1, 1))[0].id).toEqual(
      '0x06bd1af522f43bc270203e96a019b4779195b870'
    );
  });
  it('should use pools correctly', async () => {
    const { result, waitForValueToChange } = renderHook(() =>
      usePools({ chainId: O.some(1), first: 0 })
    );
    await waitForValueToChange(() => result.current.pools);

    const length = pipe(
      result.current.pools,
      O.map((r) => R.size(r))
    );
    expect(length).toMatchInlineSnapshot(`
      Object {
        "_tag": "Some",
        "value": 901,
      }
    `);
  });

  it('should trade', async () => {
    const pools = O.some(groupBySymbol(defaultPools));
    const { result, waitForValueToChange, waitForNextUpdate } = renderHook(
      () => ({
        ...useUniswap({
          ...useUniswapProps,
          pools: pools,
        }),
        quoter: useQuoter({
          provider: useUniswapProps.provider,
          chainId: O.some(chainId),
        }),
      })
    );

    await waitForValueToChange(() => result.current.pool, { timeout: 10000 });
    await waitForValueToChange(() => result.current.quoter);
    await pipe(
      sequenceT(O.Apply)(
        pools,
        result.current.pool,
        result.current.tokenAUniswap,
        result.current.tokenBUniswap,
        useUniswapProps.provider,
        result.current.quoter
      ),
      O.map(([pools, p, tokenA, tokenB, provider, q]) => {
        return [
          /*  pipe(
            pools,
            R.toArray,
            A.map((p) => createPoolFromSubgrap(1)(p[1][0]))
          ), */
          /*      pools, */
          tokenA,
          tokenB,
          provider,
          pipe(
            pools,
            R.toArray,
            A.map((p) => p[1]),
            A.flatten,
            A.map((p) => [p, createPoolContract(provider)(p.id)] as const)
          ),
          q,
        ] as const;
      }),
      TO.fromOption,
      TO.chain(([tokenA, tokenB, p, poolsAndContracts, q]) => async () => {
        const sortTickByIndex = pipe(
          N.Ord,
          contramap((t: Tick) => t.index)
        );
        const uniswapPools = await pipe(
          poolsAndContracts,
          A.map(([p, c]) => async () => {
            const ticks = pipe(
              p.ticks,
              A.map((t) => {
                return new Tick({
                  liquidityGross: t.liquidityGross,
                  liquidityNet: t.liquidityNet,
                  index: Number(t.id.split('#')[1]),
                });
              }),
              A.sort(sortTickByIndex)
            );
            try {
              return createPoolFromSubgrap(1)(p, await c.tickSpacing(), []);
            } catch (e) {
              console.log(e);
              console.log('invalid pool for', p.token0.symbol, p.token1.symbol);
            }
          }),
          T.sequenceArray,
          T.map((pools) =>
            pipe(
              [...pools],
              A.filter((t) => t != null)
            )
          )
        )();
        return O.some([uniswapPools, tokenA, tokenB, p, q] as const);
      }),
      TO.chain(([pools, tokenA, tokenB, p, q]) => async () => {
        return O.some([pools, tokenA, tokenB, p, q] as const);
      }),
      TO.chain(([pools, tokenA, tokenB, p, q]) => async () => {
        console.log(tokenA, tokenB);

        const filteredPools = pipe(
          pools,
          A.filter((p) => p.involvesToken(tokenA) || p.involvesToken(tokenB))
        );
        const routes = new Route(filteredPools, tokenB, tokenA);
        console.log('here', routes.tokenPath);
        /*     const trade = await Trade.bestTradeExactIn(
          pools,
          CurrencyAmount.fromRawAmount(tokenA, 1),
          tokenB,
          { maxNumResults: 1, maxHops: 3 }
        );

        console.log(trade); */
        return O.some(pools);
      })
    )();
  });
  /*   it('should use uniswap correctly', async () => {
    const pools = O.some(groupBySymbol(defaultPools));
    const { result, waitForValueToChange, waitForNextUpdate } = renderHook(() =>
      useUniswap({
        ...useUniswapProps,
        pools: pools,
      })
    );

    await waitForValueToChange(() => result.current.pool, { timeout: 10000 });

    await pipe(
      sequenceT(O.Apply)(
        result.current.pool,
        result.current.tokenAUniswap,
        result.current.tokenBUniswap
      ),
      O.map(([pool]) => {
        return pool;
      }),
      TO.fromOption,
      TO.chain((p) => async () => {
        console.log(await result.current.getTokenPrice(tokenA));
        console.log(await result.current.getTokenPrice(tokenB));

        return O.some(p);
      })
    )();
  }); */
  it('should use uniwap token correctly', async () => {
    const tokens = getMockTokens();
    const busd = pipe(tokens, R.lookup('WETH'));
    const dai = pipe(tokens, R.lookup('DAI'));
    const { result } = renderHook(() =>
      useUniswapTokens({
        tokenA: busd,
        tokenB: dai,
      })
    );
    const formatSnapshot = (token: O.Option<Token>) =>
      pipe(
        token,
        O.map(({ symbol, address, chainId, decimals }) => ({
          symbol,
          address,
          chainId,
          decimals,
        }))
      );
    expect(formatSnapshot(result.current.tokenA)).toMatchInlineSnapshot(`
      Object {
        "_tag": "Some",
        "value": Object {
          "address": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
          "chainId": 1,
          "decimals": 18,
          "symbol": "WETH",
        },
      }
    `);
    expect(formatSnapshot(result.current.tokenB)).toMatchInlineSnapshot(`
      Object {
        "_tag": "Some",
        "value": Object {
          "address": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
          "chainId": 1,
          "decimals": 18,
          "symbol": "DAI",
        },
      }
    `);
  });
  it('should use uniwap best route correctly', async () => {
    const tokens = getMockTokens();
    const busd = pipe(tokens, R.lookup('WETH'));
    const dai = pipe(tokens, R.lookup('DAI'));
    const someChainId = O.some(chainId);
    const pools = groupBySymbol(defaultPools);
    const poolsOption = O.some(pools);
    const uniswapPools = pipe(pools, createPoolsFromSubgraph(chainId), O.some);
    const amountIn = O.some(1);
    const { result, waitForValueToChange } = renderHook(() =>
      useUniswapRoute({
        tokenIn: busd,
        tokenOut: dai,
        chainId: someChainId,
        pools: uniswapPools,
        amountIn,
      })
    );

    expect(formatSnapshot(result.current.tokenA)).toMatchInlineSnapshot(`
      Object {
        "_tag": "Some",
        "value": Object {
          "address": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
          "chainId": 1,
          "decimals": 18,
          "symbol": "WETH",
        },
      }
    `);
    expect(formatSnapshot(result.current.tokenB)).toMatchInlineSnapshot(`
      Object {
        "_tag": "Some",
        "value": Object {
          "address": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
          "chainId": 1,
          "decimals": 18,
          "symbol": "DAI",
        },
      }
    `);
  });
});
