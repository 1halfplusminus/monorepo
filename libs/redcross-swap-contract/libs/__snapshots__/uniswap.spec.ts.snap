// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Use uniswap hook should use uniwap token correctly 2`] = `
Object {
  "_tag": "Some",
  "value": undefined,
}
`;
